cmd/kubeadm/app/util/kustomize/kustomize.go

```diff
@@ -0,0 +1,181 @@
+	// to rebuild the list of patches manually
+	// TODO: make this git friendly - currently this works only for patches in local folders -
+	files, err := ioutil.ReadDir(km.kustomizeDir)
+	if err != nil {
+		return err
+	}
+
+	var paths = []string{}
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		paths = append(paths, file.Name())
+	}
+
+	// Create a loader that mimics the behavior of kubectl kustomize, inclusing support for reading from
+	// a local git repository like git@github.com:someOrg/someRepo.git or https://github.com/someOrg/someRepo?ref=someHash
+	fSys := fs.MakeRealFS()
+	ldr, err := loader.NewLoader(km.kustomizeDir, fSys)
+	if err != nil {
+		return err
+	}
+	defer ldr.Cleanup()
+
+	// read all the kustomizations and build the UnstructuredSlice
+	us, err := NewUnstructuredSliceFromFiles(ldr, paths)
+	if err != nil {
+		return err
+	}
+
+	km.us = us
+	return nil
+}
+
+// Kustomize apply a set of patches to a resource.
+// Portions of the kustomize logic in this function are taken from the kubernetes-sigs/kind project
+func (km *Manager) Kustomize(res []byte) ([]byte, error) {
+	// create a loader that mimics the behavior of kubectl kustomize
+	// and converts the resource into a UnstructuredSlice
+	// Nb. in kubeadm we are controlling resource generation, and so we
+	// we are expecting 1 object into each resource, eg. the static pod.
+	// Nevertheless, this code is ready for more than one object per resource
+	resList, err := NewUnstructuredSliceFromBytes(res)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a list of resource and corresponding patches
+	var resources, patches UnstructuredSlice
+	for _, r := range resList {
+		resources = append(resources, r)
+
+		resourcePatches := km.us.FilterResource(r.GroupVersionKind(), r.GetNamespace(), r.GetName())
+		patches = append(patches, resourcePatches...)
+	}
+
+	fmt.Printf("[kustomize] Applying %d patches\n", len(patches))
+
+	// if there are no patches, for the target resources, exit
+	if len(patches) == 0 {
+		return res, nil
+	}
+
+	// create an in memory fs to use for the kustomization
+	memFS := fs.MakeFakeFS()
+
+	var kustomization bytes.Buffer
+	fakeDir := "/"
+	// for Windows we need this to be a drive because kustomize uses filepath.Abs()
+	// which will add a drive letter if there is none. which drive letter is
+	// unimportant as the path is on the fake filesystem anyhow
+	if runtime.GOOS == "windows" {
+		fakeDir = `C:\`
+	}
+
+	// write resources and patches to the in memory fs, generate the kustomization.yaml
+	// that ties everything together
+	kustomization.WriteString("resources:\n")
+	for i, r := range resources {
+		b, err := r.MarshalJSON()
+		if err != nil {
+			return nil, err
+		}
+
+		name := fmt.Sprintf("resource-%d.json", i)
+		_ = memFS.WriteFile(filepath.Join(fakeDir, name), b)
+		fmt.Fprintf(&kustomization, " - %s\n", name)
+	}
+
+	kustomization.WriteString("patches:\n")
```

Comment:
I think it's fine for now but would a template look more clear?

Reply:
I'm working on a follow-up PR for overcoming known limitations, and I will fix this in that PR

Reply:
We already import some Kustomize packages, importing the types package, filling in a struct and marshaling it to YAML looks like the cleanest way to me.